src/dev.rs

```diff
@@ -0,0 +1,293 @@
+//! Tools for developing circuits.
+
+use ff::Field;
+
+use crate::{
+    arithmetic::{FieldExt, Group},
+    plonk::{Any, Assignment, Circuit, Column, ConstraintSystem, Error},
+    poly::{EvaluationDomain, LagrangeCoeff, Polynomial},
+};
+
+#[derive(Debug)]
+struct Cell(usize, usize);
+
+/// The reasons why a particular circuit is not satisfied.
+#[derive(Debug, PartialEq)]
+pub enum VerifyFailure {
+    /// A gate was not satisfied for a particular row.
+    Gate { gate_index: usize, row: usize },
+    /// A lookup input did not exist in its corresponding table.
+    Lookup { lookup_index: usize, row: usize },
+    /// A permutation did not preserve the original value of a cell.
+    Permutation {
+        perm_index: usize,
+        column: usize,
+        row: usize,
```

Comment:
If the size is correct then there must be *both* some value in the input that was included too many times, and some value in the table that wasn't included enough times. Which one are we reporting? Should we report both?

Reply:
I realized after I wrote this that we are checking whether a *specific* permutation is correct, when we report this error. Can we make this clearer in the doc?